### ISSUE TITLE
[readability] establish a set of code format rules

### ISSUE THREAD

**Issue by 2bndy5** [Mmessiter/LockDownRadioControl#4]
I have trouble tracking of scope of source code when reviewing it. There are tools that can automate this, namely clang-tidy or clang-format, but we need to agree on the defined rules that these clang tools use to fix or warn about inconsistencies.

In general, I would suggest the following rules
1. 4-space indents (spaces not tabs)
2. `{` placement is on its own line in the file's global scope. However, `{` can be placed at the end of a line that starts a new code block. `}` should not be followed by any other code.
   For example:
   ```cpp
   void func()
   { // <-- file's global scope
       uint8_t var = 0;
       if (!var) { // <-- new block  (not at global scope)
           var++;
       }
       else { // <-- new block  (not at global scope)
           var--;
       }
   }
   ```
   curly-brackets may not be required for 1 line code blocks. The above snippet is just an example.
3. whitespacing
   - operators should be surrounded by 1 space.
   - commas, should be followed by 1 space.
   - opening curly-brackets should be preceeded by 1 space if not at the start of new line. (see example snippet above)
4. pre-processor commands (code that begins with `#`) follow its own scoping indentation while the rest of the code has its own scoping indentation.
   ```cpp
   void func()
   {
   #ifdef DEBUG
       Serial.println("debug output");
       #ifdef FEATURE_TOGGLE
       Serial.println("extra debug info");
       #endif // defined FEATURE_TOGGLE
    #endif // defined DEBUG
   }
   ```
   Notice the comments following the `#endif` statements; this would be greatly intuitive for multi-line encapsulated blocks. This shouldn't be required for the above example snippet since the code blocks are 1 line long.

Adding a documenting comment for function/variable definitions/declarations should be a separate issue, but it would be very helpful for onlookers like myself.

**Comment by Mmessiter** [comment]
Yes - thank you. I will try to improve this aspect, today. As you see - there's such a lot I simply don't know, yet. 
I was self-taught in 8086 assembler in the 1980s. Now I'm self-taught in C++ these last 12 months or so. It's really great that I now have really good advice from which to learn! 
(...In 1968 I learned oboe at the Paris Conservatoire of music. That seems like another planet now. 😁)

**Comment by 2bndy5** [comment]
I assigned myself this issue because it is something that you only need to be mindful of. I can fix this very easily and setup the repo to notify you when new changes don't conform to the agreed style rules.

We only need to discuss and agree upon the formatting rules in this thread (open or closed).

**Comment by Mmessiter** [comment]
As this is new to me, I'll happily agree to whichever style you propose. I'll investigate this today. 
These's much to do - docs, code, etc - so I might need a little time. But I'll do it!

**Comment by Mmessiter** [comment]
... I will try hard to follow your code indent style suggestions. When I'm developing new code it's quite likely that in my pressured mind style may take second place to sheer logic of execution - so I apologise in advance for unintentional departures from agreed style. These can be fixed when it's working!

**Comment by Mmessiter** [comment]
I have been trying to install clang tidy and clang format￼ but so far without success! What is the simplest way to run these tools?￼￼

**Comment by 2bndy5** [comment]
there are many... [PIO has its own integration](https://docs.platformio.org/en/latest/plus/check-tools/clang-tidy.html), but it seems a bit complex to set up. I normally install from the [LLVM repo's release page's downloads](https://github.com/llvm/llvm-project/releases/tag/llvmorg-12.0.1) for windows. If I'm on a Linux OS, then I use `apt install clang-tidy`. Unfortunately for you, I'm very anti-apple products, so I cannot give empirical advise in that situation.

**Comment by Mmessiter** [comment]
I'll try to discover how to do it!

**Comment by 2bndy5** [comment]
I ran across something from the LLVM docs about xcode automatically using clang as its compiler, but it didn't mention if clang-tidy or clang-format are included/installed with the compiler.

To use clang-tidy, one would just execute (from the repo's root folder as an example)
```shell
clang-tidy --checks=* ReceiverCode/main.cpp
```

**Comment by Mmessiter** [comment]
I wish I were in San Francisco! I'll try to get it working more remotely...

**Comment by 2bndy5** [comment]
Not sure how it is in London, but its still a political sh*t show over here. I wish it was warmer in Canada...

**Comment by Mmessiter** [comment]
Politics here in the uk are well worth avoiding. All involved obviously have only self interest at heart, as usual. And almost all are both idiots and pretenders. It's clear that only an idiot would want to be in government- so that's who we have! So I like to keep a low profile, write code, fly helicopters, and play oboe while awaiting a really good AI system that has more than self interest in mind and can run things better! 😉

**Comment by 2bndy5** [comment]
I started a new branch called "run-clang-format" in which I added a .clang-format file (something I've been designing for the RF24 repo) and executed clang-format on all remaining source code in the TransmitterCode and ReceiverCode directories. 

For the record, I have clang-format installed (and added to my system's PATH environment variable). Additionally, I use VSCode with the [clang-format extension](https://marketplace.visualstudio.com/items?itemName=xaver.clang-format) installed in VSCode. This way I only need to adjust the settings in the _.clang-format_ file(s) (clang-format will look for the closest relative _.clang-format_ file when its `--style` option is set to `"file"`), then right-click anywhere within the source code and select "Format Document". Personally, I prefer to use Webkit style as a fallback for when no _.clang-format_ file is present.

**Comment by Mmessiter** [comment]
You read my mind! I was about to suggest exactly that! 

I will look with interest at this branch and confirm that the code still works the same. Assuming it does perhaps it should become the main area rather than a branch ?

**Comment by 2bndy5** [comment]
> Assuming it does perhaps it should become the main area rather than a branch ?

Answering this question is exactly what PRs are most useful for.

**Comment by Mmessiter** [comment]
It looks very good - if a little unfamiliar! 
(I'd become used to the old look of course. But that's not important.)
I'm not yet up and into my work zone so I'll double check - in a few hours' time! It's still only 06.43...
